packages/vx-scale/src/scales/linear.ts

```diff
@@ -1,35 +1,35 @@
 import { scaleLinear } from 'd3-scale';
+import { Value } from '../types/Base';
+import { PickScaleConfigWithoutType } from '../types/ScaleConfig';
+import { PickD3Scale } from '../types/Scale';
+import applyInterpolate from '../mixins/applyInterpolate';
+import applyRound from '../mixins/applyRound';
+import applyZero from '../mixins/applyZero';
 
-export type LinearConfig<Output> = {
-  /** Sets the input values of the scale, which are numbers for a linear scale. */
-  domain?: number[];
-  /** Sets the output values of the scale. */
-  range?: Output[];
-  /** Sets the output values of the scale while setting its interpolator to round. If the elements are not numbers, they will be coerced to numbers. */
-  rangeRound?: number[];
-  /** Extends the domain so that it starts and ends on nice round values. */
-  nice?: boolean;
-  /** Whether the scale should clamp values to within the range. */
-  clamp?: boolean;
-};
+export function updateLinearScale<Output extends Value = Value>(
+  scale: PickD3Scale<'linear', Output>,
+  config: PickScaleConfigWithoutType<'linear', Output>,
+) {
+  const { domain, range, clamp = true, nice = true } = config;
 
-export default function linearScale<Output>({
-  range,
-  rangeRound,
-  domain,
-  nice = false,
-  clamp = false,
-}: LinearConfig<Output>) {
-  const scale = scaleLinear<Output>();
-
-  if (range) scale.range(range);
-  if (rangeRound) scale.rangeRound(rangeRound);
   if (domain) scale.domain(domain);
   if (nice) scale.nice();
-  if (clamp) scale.clamp(true);
+  if (range) scale.range(range);
 
+  scale.clamp(clamp);
+  applyInterpolate(scale, config);
+  applyRound(scale, config);
+  applyZero(scale, config);
```

Comment:
Since these are mutable functions, is call order significant here? Does `applyInterpolate()` have to be called before `applyRound()`? If so, we might want to think of a way to enforce this. I don't expect it to a problem as it is lib code, but could be an easily introduced bug that could slip through a review.

Reply:
> Since these are mutable functions, is call order significant here?

The ones that really have to be in order are `domain > nice > zero`.

> Does `applyInterpolate()` have to be called before `applyRound()`? 

Not necessary. They should be mutually exclusive. `applyInterpolate()` is in fact `applyColorInterpolate`. (This is a new functionality for `vx`. I could also drop it. Previously I don't think `vx` was accepting `interpolate`.)
Users can either do rounding which will set `interpolateRound` 

when a scale output is number.
```ts
{ round: true }
```

or 

when a scale output is color and they want to set color interpolation to HCL
```
{ interpolate: 'hcl' }
```

It is still not super clean and the typing may allow something like this to be specified. Although both effect cannot happen at the same time.
```
{ interpolate: 'hcl', round: true }
```

### Alternatives

1. I could drop the `interpolate` feature for now. 
2. Stricter typing to allow only either `{ interpolate }` or `{ round }`. Can be complicated. 
3. Make `{ round: true }` become `{ interpolate: 'round' }` but will be breaking change and more different from D3 interface.

Reply:
One idea instead of proxy (can copy & paste in browser console):
 
```js
const log = (str) => (scale, config) => console.log(str);

const operations = {
  domain: log('domain'),
  nice: log('nice'),
  interpolate: log('interpolate'),
  round: (scale, config) => config.interpolate ? console.warn(`can't round and interpolate`) : console.log('round'),
  zero: log('zero')
};

const order = Object.keys(operations);
const intersection = (a,b) => a.filter(v => b.includes(v));

function scaleOperator(...ops) {
  const orderedOps = intersection(order, ops);
  return (scale, config) => {
    return orderedOps.forEach(op => operations[op](scale, config))
  }
}

// order doesn't matter
const applyOperator1 = scaleOperator('round', 'zero', 'domain', 'nice');
const applyOperator2 = scaleOperator('round', 'zero', 'domain', 'interpolate');

// in createScale({ scale, config })
applyOperator1(/**scale*/ {}, /**config*/ {});
applyOperator2(/**scale*/ {}, /**config*/ { round: true, interpolate: true });
```
```
output 1: 
=> domain
=> nice
=> round
=> zero

output 2: 
=> domain
=> interpolate
=> WARN: can't round and interpolate
=> zero
```

Reply:
I refactor with this idea. The individual scale file now looks very clean.